### PR TITLE
Add containers.conf read-only flag support 

### PR DESCRIPTION
**Pull request by rhatdan**
If you are running temporary containers within podman play kube
we should really be running these in read-only mode. For automotive
they plan on running all of their containers in read-only temporal
mode. Adding this option guarantees that the container image is not
being modified during the running of the container.

The containers can only write to tmpfs mounted directories.

Signed-off-by: Daniel J Walsh <dwalsh@redhat.com>

### CODE REVIEW

**Comment by rhatdan** [comment]
@alexlarsson @ygalblum PTAL
@containers/podman-maintainers PTAL

File: docs/source/markdown/options/read-only.md [hunk 1]
```diff
@@ -1,11 +1,11 @@
 ####> This option file is used in:
-####>   podman create, run
+####>   podman create, kube play, run
 ####> If you edit this file, make sure your changes
 ####> are applicable to all of those.
```
Comment (TomSweeneyRedHat) [inline docs/source/markdown/options/read-only.md:4]: Suggest "those commands."

File: docs/source/markdown/options/read-only.md [hunk 2]
```diff
@@ -1,11 +1,11 @@
...
 #### **--read-only**
 
 Mount the container's root filesystem as read-only.
 
-By default a container will have its root filesystem writable allowing processes
-to write files anywhere. By specifying the **--read-only** flag, the container will have
-its root filesystem mounted as read-only prohibiting any writes.
+By default containers have their root filesystem writable allowing processes
```
Comment (TomSweeneyRedHat) [inline docs/source/markdown/options/read-only.md:9]:
```suggestion
By default, containers have their root filesystem writable, allowing processes
```

File: docs/source/markdown/options/read-only.md [hunk 3]
```diff
@@ -1,11 +1,11 @@
...
+to write files anywhere. By specifying the **--read-only** flag, containers
+will have root filesystems mounted as read-only prohibiting any writes.
```
Comment (TomSweeneyRedHat) [inline docs/source/markdown/options/read-only.md:11]:
```suggestion
will have root filesystems mounted as read-only, prohibiting any writes.
```

File: docs/source/markdown/options/read-only.md (hunk 1 above)
Comment (rhatdan) [reply docs/source/markdown/options/read-only.md:4]: These are auto-generated.

**Comment by Luap99** [comment]
@rhatdan Can you squash these PR comment commits please into your actual one, they increase the commit count for no reason and if I browse the commit log they provide no clue about the change.

**Review by Luap99 — COMMENTED** [review]
Should we just use
```
securityContext:
      readOnlyRootFilesystem: true
```
https://kubernetes.io/docs/reference/kubernetes-api/workload-resources/pod-v1/
in the yaml instead of adding this as flag?

**Comment by rhatdan** [comment]
I think we might want to use both.  The goal is to allow automotive to force Pods to be read-only, potentially similar in quadlet. 
It looks like that option does not currently work in containers either.

**Comment by rhatdan** [comment]
Actually my code breaks this use case.

**Comment by alexlarsson** [comment]
Since we already have an yaml options for this, I think we should use that, and then add support for changing its default value. Then quadlet can e.g. do `--default-container=securityContext.readOnlyRootFilesystem=true` or something like that. This seems more flexible than adding a bunch of options with unclear semantics in cases where e.g. some containers in the pod specify the field and some don't.

**Comment by alexlarsson** [comment]
Similarly, we could have `--default-annotation=`

**Comment by ygalblum** [comment]
As @Luap99 and @alexlarsson wrote, since there is a field for it in the K8S schema, `podman kube play` should use it and not add another flag for it. Having said that, as @alexlarsson wrote, we would like to be able to set a different default value for when the value is not set.
As for the suggestion regarding `--default-container` since `container` is a structure, maybe the argument should be a JSON. This way the user does not have to repeat the parameter if she wished to set more than one default value. Moreover, if it's a JSON, we should consider allowing it to be passed as a path to a file (maybe with the a prefix of `@` like in Ansible)

**Comment by rhatdan** [comment]
To me that is a bit of a contradiction to have the ability to override just in quadlet but not from the command line.  But I am not wed to having this option.  Most of the framework of this PR is still required to allow quadlet to set that default.

**Comment by ygalblum** [comment]
Maybe I wasn't clear. I didn't mean for these flags to be `quadlet` specific, nor do I see how it can be done. I think these should be `podman kube play` flags that callers can use. One such caller is `quadlet` but any user can pass them.

File: cmd/podman/kube/play.go [hunk 4]
```diff
@@ -137,6 +138,7 @@ func playFlags(cmd *cobra.Command) {
 	flags.BoolVar(&playOptions.NoHosts, "no-hosts", false, "Do not create /etc/hosts within the pod's containers, instead use the version from the image")
 	flags.BoolVarP(&playOptions.Quiet, "quiet", "q", false, "Suppress output information when pulling images")
 	flags.BoolVar(&playOptions.TLSVerifyCLI, "tls-verify", true, "Require HTTPS and verify certificates when contacting registries")
+	flags.BoolVar(&playOptions.ReadOnlyCLI, "read-only", false, "Make all containers root filesystem read-only")
```
Comment (vrothberg) [inline cmd/podman/kube/play.go:141]:
```suggestion
	flags.BoolVar(&playOptions.ReadOnlyCLI, "read-only", false, "Make all containers' root filesystems read-only")
```

File: docs/source/markdown/options/add-host.md [hunk 5]
```diff
@@ -1,7 +1,7 @@
 ####> This option file is used in:
 ####>   podman build, create, pod create, run
 ####> If you edit this file, make sure your changes
-####> are applicable to all of those.
+####> are applicable to all of those files.
```
Comment (vrothberg) [inline docs/source/markdown/options/add-host.md:4]: This change should be broken out into another commit.  It's causing a lot of noise.

File: pkg/specgen/generate/kube/kube.go [hunk 6]
```diff
@@ -444,6 +447,10 @@ func ToSpecGen(ctx context.Context, opts *CtrSpecGenOptions) (*specgen.SpecGener
 		}
 	}
 
+	if ro := opts.ReadOnly; ro != itypes.OptionalBoolUndefined {
+		s.ReadOnlyFilesystem = (ro == itypes.OptionalBoolTrue)
+	}
```
Comment (vrothberg) [inline pkg/specgen/generate/kube/kube.go:452]: Should be another commit IMO and tested.

File: pkg/domain/infra/tunnel/kube.go [hunk 7]
```diff
@@ -66,6 +66,9 @@ func (ic *ContainerEngine) PlayKube(ctx context.Context, body io.Reader, opts en
 		options.WithAnnotations(opts.Annotations)
 	}
 	options.WithNoHosts(opts.NoHosts).WithUserns(opts.Userns)
+	if ro := opts.ReadOnly; ro != types.OptionalBoolUndefined {
+		options.WithReadOnly(ro == types.OptionalBoolTrue)
```
Comment (vrothberg) [inline pkg/domain/infra/tunnel/kube.go:70]: I don't see plumbing for it on the server side.

**Comment by ygalblum** [comment]
Considering the fact that the Pod manifest already has a field (at the container level) to control the "read-only" attribute, do we still want to add this flag?  
If we add the flag, do we disregard the value in the manifest? If we also look at the manifest, what takes precedence?

**Comment by rhatdan** [comment]
The manifest is primary, but if the user overrides the flag, then the flag takes precedence.  

podman kube play foobar.yaml # Yaml wins
podman kube play --readonly foobar.yaml # Read Only
podman kube play --readonly=false foobar.yaml # Read Write

**Comment by rhatdan** [comment]
@vrothberg @alexlarsson @mheon @Luap99 PTAL

Should we merge or should I remove the CLI overrides?

**Comment by Luap99** [comment]
I would prefer to remove the cli option. Everything that can be set in yaml should be set there, if we start adding these overwrites we might end up adding all podman run options which I like to avoid as much as possible.

If quadlet wants to change some defaults then we should come up with a generic solution for that, adding a bunch of flags for that seems unnecessary.

**Comment by rhatdan** [comment]
But the issues is Quadlet is just executing a podman command under the covers. So if quadlet wants to define a
read-only option when executing kube.yaml files it can't unless the option is exposed in podman. Optionally it could attempt to edit the kubernetes.yaml file on the fly, but this would be difficult when YAML has multiple pods and containers.

**Comment by vrothberg** [comment]
> But the issues is Quadlet is just executing a podman command under the covers. So if quadlet wants to define a read-only option when executing kube.yaml files it can't unless the option is exposed in podman. Optionally it could attempt to edit the kubernetes.yaml file on the fly, but this would be difficult when YAML has multiple pods and containers.

Yes, in that case, the hands are tied.  Is Quadlet setting such a default a reasonable use case?  Since it's exposed in the K8s YAML and not something Podman-specific, it seems dangerous trying to be smarter than the user.

I do not feel strongly about this particular case but if there's trend to add many container-create options to `kube play`, we may be going down the wrong path.  The nice thing about K8s YAML is that the developer decides how to run the container.

**Comment by rhatdan** [comment]
Problem with that is the developer is often lacking in security concerns and is more worried about working then necessarily security or application hygiene, and in some cases performance of the overall system.

**Comment by rhatdan** [comment]
@containers/podman-maintainers PTAL
@alexlarsson @ygalblum PTAL

File: cmd/podman/containers/create.go [hunk 8]
```diff
@@ -409,6 +409,8 @@ func createPodIfNecessary(cmd *cobra.Command, s *specgen.SpecGenerator, netOpts
 	infraOpts := entities.NewInfraContainerCreateOptions()
 	infraOpts.Net = netOpts
 	infraOpts.Quiet = true
+	infraOpts.ReadOnly = true
+	infraOpts.ReadOnlyTmpFS = false
```
Comment (vrothberg) [inline cmd/podman/containers/create.go:413]: Could we set it to `true`?  We do it for the service-container which is essentially the same as an infra-container.

File: pkg/domain/infra/abi/play.go [hunk 9]
```diff
@@ -605,6 +609,15 @@ func (ic *ContainerEngine) playKubePod(ctx context.Context, podName string, podY
 		}
 	}
 
+	cfg, err := ic.Libpod.GetConfigNoCopy()
+	if err != nil {
+		return nil, nil, err
+	}
+	var readOnly types.OptionalBool
+	if cfg.Containers.ReadOnly {
+		readOnly = types.NewOptionalBool(cfg.Containers.ReadOnly)
```
Comment (vrothberg) [inline pkg/domain/infra/abi/play.go:618]: decl-setting it directly `readOnly := types.NewOptionalBool(cfg.Containers.ReadOnly)` without the if statement would probably be faster.

File: pkg/specgen/generate/storage.go [hunk 10]
```diff
@@ -427,3 +432,31 @@ func InitFSMounts(mounts []spec.Mount) error {
 	}
 	return nil
 }
+
+func addReadOnlyMounts(mounts map[string]spec.Mount, volumes []*specgen.NamedVolume) map[string]spec.Mount {
```
Comment (vrothberg) [inline pkg/specgen/generate/storage.go:436]: The function name confuses me. "ReadOnly" but we mount "rw".

File: pkg/specgenutil/specgen.go [hunk 11]
```diff
@@ -593,10 +593,8 @@ func FillOutSpecGen(s *specgen.SpecGenerator, c *entities.ContainerCreateOptions
 		s.DependencyContainers = c.Requires
 	}
 
-	// TODO
-	// outside of specgen and oci though
-	// defaults to true, check spec/storage
-	// s.readonly = c.ReadOnlyTmpFS
+	s.ReadWriteTmpfs = c.ReadOnly && c.ReadOnlyTmpFS
```
Comment (vrothberg) [inline pkg/specgenutil/specgen.go:596]: Same here.  I didn't read the entire code but `ReadWrite` is set when the container is configured to be read only?

**Review by vrothberg — COMMENTED** [review]
Minor nits.  Some code comments may help resolve my confusion.

File: cmd/podman/containers/create.go (hunk 8 above)
Comment (vrothberg) [reply cmd/podman/containers/create.go]: The infra container sets it to `true` and I wonder why they are configured differently.

Comment (rhatdan) [reply cmd/podman/containers/create.go]: Switched that one to false also.

File: pkg/domain/infra/abi/play.go (hunk 9 above)
Comment (rhatdan) [reply pkg/domain/infra/abi/play.go:619]: Since false is the default, we want to leave as an optionalbool without being set, that way the settings from the kube yaml take over.  If I set this OptionalBool to False, then it ignores the kube.yaml settings.

File: pkg/specgenutil/specgen.go (hunk 11 above)
Comment (rhatdan) [reply pkg/specgenutil/specgen.go]: Switched internal Variables to ReadWriteTmpFS.

**Comment by rhatdan** [comment]
@containers/podman-maintainers PTAL
@vrothberg PTAL

File: pkg/domain/infra/abi/play.go (hunk 9 above)
Comment (vrothberg) [reply pkg/domain/infra/abi/play.go:619]: Ah, that makes sense, thanks!

File: pkg/domain/infra/abi/play.go [hunk 12]
```diff
@@ -560,6 +562,8 @@ func (ic *ContainerEngine) playKubePod(ctx context.Context, podName string, podY
 		infraImage := util.DefaultContainerConfig().Engine.InfraImage
 		infraOptions := entities.NewInfraContainerCreateOptions()
 		infraOptions.Hostname = podSpec.PodSpecGen.PodBasicConfig.Hostname
+		infraOptions.ReadOnly = true
+		infraOptions.ReadWriteTmpFS = false
```
Comment (vrothberg) [inline pkg/domain/infra/abi/play.go:566]: To be sure: the service-container sets it `true`, the infra to `false`.  Is that intentional? Both execute the same work.

**Review by vrothberg — COMMENTED** [review]
One question but LGTM

File: pkg/domain/infra/abi/play.go (hunk 12 above)
Comment (rhatdan) [reply pkg/domain/infra/abi/play.go:566]: The both should set it False now, but yes the same logic applies.

**Comment by rhatdan** [comment]
@giuseppe @alexlarsson @ygalblum @flouthoc PTAL

File: pkg/domain/infra/abi/play.go (hunk 12 above)
Comment (vrothberg) [reply pkg/domain/infra/abi/play.go:566]: The service container sets it to `true` though (https://github.com/containers/podman/pull/16545/files#diff-d3fb054de60fb7550ae0554fc1c99ac63b8434f960b49b5b7f0352f1d0fcb8f0R72).

Comment (rhatdan) [reply pkg/domain/infra/abi/play.go:566]: Shoot.

File: pkg/specgen/generate/storage.go [hunk 13]
```diff
@@ -427,3 +432,31 @@ func InitFSMounts(mounts []spec.Mount) error {
 	}
 	return nil
 }
+
+func addReadWriteTmpfsMounts(mounts map[string]spec.Mount, volumes []*specgen.NamedVolume) map[string]spec.Mount {
+	readonlyTmpfs := []string{"/tmp", "/var/tmp", "/run"}
+	options := []string{"rw", "rprivate", "nosuid", "nodev", "tmpcopyup"}
+	for _, dest := range readonlyTmpfs {
+		for _, m := range mounts {
```
Comment (ygalblum) [inline pkg/specgen/generate/storage.go:440]: From the last line of the outer loop, I can see that the `dest` field is also used as the key in the `mounts` map. If so, can you replace this loop with checking if `dest` exists in the map?

File: pkg/specgen/specgen.go [hunk 14]
```diff
@@ -384,6 +384,10 @@ type ContainerSecurityConfig struct {
 	// ReadOnlyFilesystem indicates that everything will be mounted
 	// as read-only
 	ReadOnlyFilesystem bool `json:"read_only_filesystem,omitempty"`
+	// ReadWriteTmpfs indicates that when running with a ReadOnlyFilesystem
+	// mount temporary file systems
+	ReadWriteTmpfs bool `json:"read_wrie_tmpfs,omitempty"`
```
Comment (ygalblum) [inline pkg/specgen/specgen.go:389]: Typo in the json name, should be `read_write_tmpfs`

File: pkg/specgenutil/specgen.go [hunk 15]
```diff
@@ -592,10 +592,10 @@ func FillOutSpecGen(s *specgen.SpecGenerator, c *entities.ContainerCreateOptions
 		s.DependencyContainers = c.Requires
 	}
 
-	// TODO
-	// outside of specgen and oci though
-	// defaults to true, check spec/storage
-	// s.readonly = c.ReadOnlyTmpFS
+	// Only add ReadWrite tmpfs mounts iff the container is ReadOnly and
+	// the user did not disable the --read-only-tmpfs flag.
```
Comment (ygalblum) [inline pkg/specgenutil/specgen.go:596]: This comment is based on the assumption that `ReadWriteTmpFS'`s default is `true`. But, the default value is defined somewhere else (in `cmd`). If for some reason, the default is changed, no one will know to change this comment and it will left with wrong information. Maybe it's better to leave the `user` out of this commnet

File: pkg/specgen/generate/storage.go (hunk 13 above)
Comment (rhatdan) [reply pkg/specgen/generate/storage.go:440]: Nice catch.

File: pkg/specgen/specgen.go (hunk 14 above)
Comment (rhatdan) [reply pkg/specgen/specgen.go:389]: Yikes.

File: pkg/specgenutil/specgen.go (hunk 15 above)
Comment (rhatdan) [reply pkg/specgenutil/specgen.go]: I want future maintainers to understand what this logic is doing, so I attempted to improve the comment.

**Comment by vrothberg** [comment]
> pkg/specgenutil/specgen.go:595:58: `contianer` is a misspelling of `container` (misspell)

Happens to me all the time as well.

File: pkg/specgenutil/specgen.go [hunk 16]
```diff
@@ -592,10 +592,11 @@ func FillOutSpecGen(s *specgen.SpecGenerator, c *entities.ContainerCreateOptions
 		s.DependencyContainers = c.Requires
 	}
 
-	// TODO
-	// outside of specgen and oci though
-	// defaults to true, check spec/storage
-	// s.readonly = c.ReadOnlyTmpFS
+	// Only add ReadWrite tmpfs mounts iff the container is container is
+	// being run ReadOnly and ReadWrite tmpfs are disabled,
```
Comment (ygalblum) [inline pkg/specgenutil/specgen.go:596]: - You have `container is` twice
- Shouldn't this be `ReadWrite tmpfs is not disabled`?

Comment (rhatdan) [reply pkg/specgenutil/specgen.go]: Yup, hopefully no more typos...

**Review by vrothberg — APPROVED** [review]
LGTM
@giuseppe PTAL

**Comment by rhatdan** [comment]
@flouthoc PTAL
@ygalblum PTAL

**Comment by ygalblum** [comment]
LGTM

**Review by flouthoc — APPROVED** [review]
LGTM
